### PR TITLE
Add cockpit-stub 

### PR DESCRIPTION
**Pull request by petervo**
Provides a limited bridge that can be used with logins mechanisms that don't tie back to a user to a real system user with a shell.

This was the best name i came up with, but i'm not really happy with it. Any other ideas?

### CODE REVIEW

**Comment by stefwalter** [comment]
After thinking about it, I would suggest:

 * ```cockpit-stub```
 * Lives in ```/usr/libexec```

The reason it's called a 'stub' is because it's not a proper bridge into the system. It's a stub that provides enough channels for **certain** components to be reused in a non-system setting. This is not the main focus of Cockpit, but rather a reuse case.

In addition, the cockpit-stub could be built into something that's functional for the panel case. Where we log into one system, but don't manage that system at all, just connect to other real bridges on real systems.

File: src/bridge/Makefile.am [hunk 1]
```diff
@@ -1,15 +1,39 @@
 
 noinst_LIBRARIES += \
+	libcockpit-shellless-bridge.a \
```
Comment (stefwalter) [inline src/bridge/Makefile.am]: We would call this libcockpit-stub.a

File: src/bridge/Makefile.am [hunk 2]
```diff
@@ -97,6 +117,16 @@ cockpit_polkit_SOURCES = src/bridge/cockpitpolkithelper.c
 cockpit_polkit_CFLAGS = $(COCKPIT_POLKIT_CFLAGS)
 cockpit_polkit_LDADD = libreauthorize.a $(REAUTHORIZE_LIBS) $(COCKPIT_POLKIT_LIBS)
 
+bin_PROGRAMS += cockpit-shellless-bridge
+
+cockpit_shellless_bridge_SOURCES = src/bridge/shellless-bridge.c
```
Comment (stefwalter) [inline src/bridge/Makefile.am]: This source file would be called stub.c

File: src/bridge/bridge.c [hunk 3]
```diff
@@ -60,46 +61,9 @@
...
-      g_message ("unsupported version of cockpit protocol: %" G_GINT64_FORMAT, version);
-      cockpit_transport_close (transport, "not-supported");
-    }
-}
-
-static struct {
-  const gchar *name;
-  GType (* function) (void);
-} payload_types[] = {
+CockpitPayloadType payload_types[] = {
```
Comment (stefwalter) [inline src/bridge/bridge.c]: Shouldn't this be ```static```?

File: src/bridge/cockpitbridge.c [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+#include "config.h"
+
+#include "cockpitchannel.h"
+#include "cockpitbridge.h"
+
+#include "common/cockpitjson.h"
+#include "common/cockpitunicode.h"
+#include "common/cockpittransport.h"
+
+#include <json-glib/json-glib.h>
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: Why do we need this header?

File: src/bridge/cockpitbridge.c [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include "cockpitchannel.h"
+#include "cockpitbridge.h"
+
+#include "common/cockpitjson.h"
+#include "common/cockpitunicode.h"
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: Why do we need this header?

File: src/bridge/cockpitbridge.c [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+#include "common/cockpittransport.h"
+
+#include <json-glib/json-glib.h>
+
+#include <glib/gstdio.h>
+
+#include <errno.h>
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: I don't see any use of EXXXX.

File: src/bridge/cockpitbridge.c [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+#include <stdlib.h>
+#include <string.h>
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: Could be wrong, but I don't think we need these either.

File: src/bridge/cockpitbridge.c [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+                                                         G_PARAM_CONSTRUCT_ONLY |
+                                                         G_PARAM_STATIC_STRINGS));
+}
+
+static void
+cockpit_bridge_init (CockpitBridge *self)
+{
+  /* Owns the channels */
+  self->channels = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_object_unref);
+  self->payloads = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: Aren't these static strings? Do they need to be duped here?

File: src/bridge/cockpitbridge.c [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+};
+
+G_DEFINE_TYPE (CockpitBridge, cockpit_bridge, G_TYPE_OBJECT);
+
+enum {
+  PROP_0,
+  PROP_INIT_RECEIVED,
+};
+
+typedef GType (* type_function) (void);
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: Function typedefs are typically CamelCase.

File: src/bridge/cockpitbridge.c [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+  const gchar *payload;
+
+  if (!channel_id)
+    {
+      g_warning ("Caller tried to open channel with invalid id");
+      cockpit_transport_close (transport, "protocol-error");
+    }
+  else if (g_hash_table_lookup (self->channels, channel_id))
+    {
+      g_warning ("Caller tried to reuse a channel that's already in use");
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: Lets start printing the channel id here: ```"%s: caller tried to use channel that's already in use"```

File: src/bridge/cockpitbridge.c [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+      /* This will close with "not-supported" */
+      channel_type = COCKPIT_TYPE_CHANNEL;
+
+      channel_function = g_hash_table_lookup (self->payloads, payload);
+      if (channel_function != NULL)
+        {
+          channel_type = channel_function ();
+        }
+
+      g_debug ("Running channel type %lu for %s default is %lu", channel_type, payload, COCKPIT_TYPE_CHANNEL);
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: If we want this message to be readable we should use ```g_type_name ()``` to format the types. Are you sure we need to list the default type?

File: src/bridge/cockpitbridge.c [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+  CockpitBridge *bridge  = g_object_new (COCKPIT_TYPE_BRIDGE,
+                                         "init-received", init_received,
+                                         NULL);
+
+  for (i = 0; payload_types[i].name != NULL; i++)
+    {
+      cockpit_bridge_add_payload (bridge, payload_types[i].name,
+                                  payload_types[i].function);
+    }
+  g_signal_connect (transport, "control", G_CALLBACK (on_transport_control), bridge);
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.c]: So the bridge should last until the channel is closed? If not, then we should probably comment here how the if the bridge is closed, but the transport is still open, then things will go boom.

File: src/bridge/cockpitbridge.h [hunk 13]
```diff
@@ -0,0 +1,49 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2014 Red Hat, Inc.
```
Comment (stefwalter) [inline src/bridge/cockpitbridge.h]: 2015?

File: src/bridge/shellless-bridge.c [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+          g_print ("    ");
+          offset = 4;
+        };
+
+      g_print ("%s ", payload_types[i].name);
+      offset += len + 1;
+    }
+  g_print ("\n");
+
+  g_print ("Authorization: crypt1\n");
```
Comment (stefwalter) [inline src/bridge/shellless-bridge.c]: I don't think cockpit-stub does authorization.

File: src/bridge/shellless-bridge.c [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+#endif
+
+  g_setenv ("GSETTINGS_BACKEND", "memory", TRUE);
+  g_setenv ("GIO_USE_PROXY_RESOLVER", "dummy", TRUE);
+  g_setenv ("GIO_USE_VFS", "local", TRUE);
+
+  context = g_option_context_new (NULL);
+  g_option_context_add_main_entries (context, entries, NULL);
+  g_option_context_set_description (context,
+                                    "cockpit-userless-bridge is run automatically inside of a Cockpit session. When\n"
```
Comment (stefwalter) [inline src/bridge/shellless-bridge.c]: This description needs fixing.

File: tools/cockpit.spec [hunk 16]
```diff
@@ -124,6 +124,13 @@ Requires: pcp
 %description pcp
 Cockpit support for reading PCP metrics and loading PCP archives.
 
+%package shellless-bridge
```
Comment (stefwalter) [inline tools/cockpit.spec]: Can we just build this as part of cockpit-ws and deliver it inside of that RPM?

**Comment by petervo** [comment]
Renamed and made the suggested changes.

**Pull request closed without merging**